Mathlib/Topology/ContinuousFunction/CFC.lean

```diff
@@ -0,0 +1,1111 @@
+    (g : C(spectrum R (cfc₁ R a f), R))
+    (f' : C(spectrum R a, spectrum R (cfc₁ R a f))) (hff' : ∀ x, (f' x : R) = f x) :
+    cfc₁ R a (g.comp f') = cfc₁ R (cfc₁ R a f) g := by
+  let cfc₃ : C(spectrum R (cfc₁ R a f), R) →⋆ₐ[R] A := (cfc₁ R a).comp (f'.compStarAlgHom' R R)
+  let this : CFCCoreClass R (cfc₁ R a f) :=
+    { toStarAlgHom := cfc₃
+      hom_continuous := CFCClass.hom_closedEmbedding.continuous.comp f'.continuous_comp_left
+      hom_map_X := by
+        simp only [cfc₂, StarAlgHom.coe_comp, Function.comp_apply]
+        congr 1
+        ext x
+        simp [hff' x] }
+  exact FunLike.congr_fun ((CFCCoreClass.ext_iff _ _).mp (Subsingleton.elim this _)) g
+
+theorem cfc₁₂_comp (f : C(spectrum R a, R)) [CFCCoreClass R (cfc₁ R a f)]
+    [Subsingleton (CFCCoreClass R (cfc₁ R a f))] (g : C(R, R)) :
+    cfc₁ R a (g.comp f) = cfc₂ R (cfc₁ R a f) g :=
+  let f' : C(spectrum R a, spectrum R (cfc₁ R a f)) :=
+    ⟨fun r => ⟨f r, cfc₁_map_spectrum a f ▸ Set.mem_range_self r (f := f)⟩,
+      (map_continuous f).subtype_mk _⟩
+  cfc₁_comp a f (g.restrict (spectrum R (cfc₁ R a f))) f' (fun _ => rfl)
+
+theorem cfc₂_comp (f g : C(R, R)) [hf₁ : CFCCoreClass R (cfc₂ R a f)]
+    [hf₂ : Subsingleton (CFCCoreClass R (cfc₂ R a f))] :
+    cfc₂ R a (g.comp f) = cfc₂ R (cfc₂ R a f) g := by
+  have : CFCCoreClass R (cfc₁ R a (f.restrict (spectrum R a))) := hf₁
+  have : Subsingleton (CFCCoreClass R (cfc₁ R a (f.restrict (spectrum R a)))) := hf₂
+  convert cfc₁₂_comp a (f.restrict (spectrum R a)) g
+
+lemma cfc₁₂_polynomial_comm (p : R[X]) [h : CFCCoreClass R (aeval (R := R) a p)]
+    [h' : Subsingleton (CFCCoreClass R (aeval (R := R) a p))] (f : C(R, R)) :
+    cfc₁ R a (f.comp (p.toContinuousMapOn (spectrum R a))) = cfc₂ R (aeval (R := R) a p) f := by
+  have key : (cfc₁ R a (p.toContinuousMapOn (spectrum R a))) = aeval (R := R) a p := by
+    simpa using cfc₁_map_polynomial a p
+  have : CFCCoreClass R (cfc₁ R a (p.toContinuousMapOn (spectrum R a))) := by convert h
+  have : Subsingleton (CFCCoreClass R (cfc₁ R a (p.toContinuousMapOn (spectrum R a)))) := by
+    convert h'
+  convert cfc₁₂_comp a (p.toContinuousMapOn (spectrum R a)) f
+  exact key.symm
+
+lemma cfc₂_polynomial_comm (p : R[X]) [CFCCoreClass R (aeval (R := R) a p)]
+    [Subsingleton (CFCCoreClass R (aeval (R := R) a p))] (f : C(R, R)) :
+    cfc₂ R a (f.comp (p.toContinuousMap)) = cfc₂ R (aeval (R := R) a p) f :=
+  cfc₁₂_polynomial_comm a p f
+
+lemma cfc₁₂_pow_comm (n : ℕ) [h : CFCCoreClass R (a ^ n)]
+    [h' : Subsingleton (CFCCoreClass R (a ^ n))] (f : C(R, R)) :
+    cfc₁ R a (f.comp (((X : R[X]) ^ n).toContinuousMapOn (spectrum R a))) = cfc₂ R (a ^ n) f := by
+  have key : aeval (R := R) a ((X : R[X]) ^ n) = a ^ n := by simp
+  have : CFCCoreClass R (aeval (R := R) a ((X : R[X]) ^ n)) := by convert h
+  have : Subsingleton (CFCCoreClass R (aeval (R := R) a ((X : R[X]) ^ n))) := by convert h'
+  convert cfc₂_polynomial_comm a ((X : R[X]) ^ n) f
```

Comment:
I'm a bit worried about this pattern coming up over and over. I guess the expectation is that 99% of the time we'll be working over a type for which we have these instances for e.g. all normal elements and they're therefore easy to find?